### PR TITLE
Add PLAID indexer and searcher implementations

### LINKED ISSUE #5: Add PLAID
https://dl.acm.org/doi/10.1145/3511808.3557325
https://dl.acm.org/doi/10.1145/3626772.3657856

### PR DESCRIPTION
**Pull request by RaykKretzschmar**
The changes introduce a new PLAID indexer and searcher using the fast-plaid library, enhancing the indexing and searching capabilities within the Lightning IR framework. Tests for the new functionality have been added and existing tests refactored for improved readability. The implementation includes methods for handling insufficient embeddings and ensures index configuration persistence.

Fixes #5

### CODE REVIEW

File: lightning_ir/retrieve/plaid/plaid_indexer.py [hunk 1]
```diff
@@ -67,79 +61,94 @@ def add(self, index_batch: IndexBatch, output: BiEncoderOutput) -> None:
...
         self.num_docs += len(doc_ids)
 
         self.doc_lengths.extend(doc_lengths.int().cpu().tolist())
         self.doc_ids.extend(doc_ids)
 
-    def process_embeddings(self, embeddings: torch.Tensor) -> torch.Tensor:
-        """Process embeddings before indexing.
+        if self.index is None:
+            if self._train_embeddings is None:
+                self._train_embeddings = embeddings.new_full(
```
Comment (fschlatt) [inline lightning_ir/retrieve/plaid/plaid_indexer.py]: We will be removing the train embeddings anyway (see comment above), but aggregating this way creates problems, since any documents with different number of tokens other than the first batch will be thrown out completely. The embeddings are padded with nans and any embeddings with nans are removed later.

Instead, aggregate the embeddings into a list. The example below shows how. Fast plaid converts a tensor into a list of tensors internally anyway, so we are not losing any efficiency.


```suggestion
    # in the __init__ function
    self.embeddings = []

    # in the add function
    self.embeddings.extend(embeddings.split(doc_lengths.int().cpu().tolist()))
```

File: lightning_ir/retrieve/plaid/plaid_searcher.py [hunk 2]
```diff
@@ -167,28 +54,29 @@ def search(self, output: BiEncoderOutput) -> Tuple[PackedTensor, List[List[str]]
...
 
-        # compute scores
-        doc_embeddings = self._reconstruct_doc_embeddings(candidate_idcs)
-        output.doc_embeddings = doc_embeddings
-        output = self.module.model.score(output, num_docs)
-        scores = output.scores
-        if scores is None:
-            raise ValueError("Expected scores in BiEncoderOutput")
+        scores = self.index.search(
+            queries_embeddings=output.query_embeddings.embeddings.detach(),
```
Comment (fschlatt) [inline lightning_ir/retrieve/plaid/plaid_searcher.py]: Why the detach? The embeddings should, in theory, never have gradients at this point

File: lightning_ir/retrieve/plaid/plaid_indexer.py [hunk 3]
```diff
@@ -67,79 +61,94 @@ def add(self, index_batch: IndexBatch, output: BiEncoderOutput) -> None:
...
+
+            if self._num_buffered < num_train:
+                return
+
+            train_embs = self._train_embeddings
+            if torch.isnan(train_embs).any():
+                train_embs = train_embs[~torch.isnan(train_embs).any(dim=1)]
+
+            config_path = self.index_dir / "config.json"
+            config_bytes = config_path.read_bytes() if config_path.exists() else None
```
[the remark this answers is not in the document]
Comment (RaykKretzschmar) [reply lightning_ir/retrieve/plaid/plaid_indexer.py]: Because fast-plaid handles the files in a way that automatically deletes the config.json
